File: common/src/main/java/com/common/entities/templates/Projectile.java
package com.common.entities.templates;

import com.common.Data;
import com.common.DataClasses.ImageWrapper;
import com.common.DataClasses.Position;
import com.common.entities.algorithms.PathFinding;
import com.common.entities.player.Player;
import com.common.graphics.Icons;
import com.common.listeners.Listeners;
import com.common.listeners.RefreshListener;
import com.common.map.Collisions;

import static com.common.Data.graphics;
import static com.common.Debug.Flags.Entities.Templates.PROJECTILE;
import static com.common.Debug.Levels.CORE;
import static com.common.Debug.Levels.INFORMATION;
import static com.common.Debug.debug;
import static com.common.graphics.Graphics.DECOR_LAYER;

public class Projectile implements RefreshListener {
	private Position position;
	private final int directionIndex;
	private final ImageWrapper icon = Icons.Interactive.projectile;
	private int damage = 5;

	public Projectile(Position position, Position target) {
		debug(PROJECTILE, CORE, "--- [Projectile.constructor]");

		this.position = position;
		directionIndex = PathFinding.toPoint(position, target);

		onRefresh();
	}

	@Override
	public void onRefresh() {
		debug(PROJECTILE, CORE, ">>> [Projectile.onRefresh]");

		// Checking, if projectile still can do some damage
		if (damage <= 0) {
			Listeners.removeRefreshListener(this);
			debug(PROJECTILE, CORE, "<<< [Projectile.onRefresh]");
			return;
		}

		// Checking for collision with player
		if (position.equals(Data.Player.position)) {
			Player.getDamage(damage);
			Listeners.removeRefreshListener(this);

			debug(PROJECTILE, CORE, "<<< [Projectile.onRefresh]");
			return;
		}

		// Case with collision with other things
		RefreshListener refreshListener = Listeners.getRefreshListener(position);
		if (refreshListener != null) {
			if (position.equals(Listeners.getRefreshListener(position))) {
				refreshListener.getEntityDamage(damage);
				Listeners.removeRefreshListener(this);

				debug(PROJECTILE, CORE, "<<< [Projectile.onRefresh]");
				return;
			}
		}

		position = PathFinding.getNextPosition(position, directionIndex);

		int couldMove = Collisions.checkForCollision(graphics.getTile(position));
		if (couldMove == Collisions.immovable) {
			Listeners.removeRefreshListener(this);

			debug(PROJECTILE, CORE, "<<< [Projectile.onRefresh]");
			return;
		}

		// Checking for collision with player
		if (position.equals(Data.Player.position)) {
			Player.getDamage(damage);
			Listeners.removeRefreshListener(this);

			debug(PROJECTILE, CORE, "<<< [Projectile.onRefresh]");
			return;
		}

		// Case with collision with other things
		refreshListener = Listeners.getRefreshListener(position);
		if (refreshListener != null) {
			if (position.equals(Listeners.getRefreshListener(position))) {
				refreshListener.getEntityDamage(damage);
				Listeners.removeRefreshListener(this);

				debug(PROJECTILE, CORE, "<<< [Projectile.onRefresh]");
				return;
			}
		}

		graphics.drawTile(position, icon, DECOR_LAYER);

		debug(PROJECTILE, CORE, "<<< [Projectile.onRefresh]");
	}

	@Override
	public void getEntityDamage(int damage) {
		debug(PROJECTILE, INFORMATION, "--- [Projectile.getEntityDamage]");
		if (damage > 0) {
			this.damage -= damage;
		}
	}

	@Override
	public Position getPosition() {
		debug(PROJECTILE, INFORMATION, "--- [Projectile.getPosition]");

		return position;
	}
}

File: composeApp/src/commonMain/kotlin/com/one_of_many_simons/one_button_game/entities/templates/Projectile.kt
package com.one_of_many_simons.one_button_game.entities.templates

import com.one_of_many_simons.one_button_game.Data
import com.one_of_many_simons.one_button_game.Debug.debug
import com.one_of_many_simons.one_button_game.Debug.Levels.CORE
import com.one_of_many_simons.one_button_game.Debug.Flags.Entities.Templates.PROJECTILE
import com.one_of_many_simons.one_button_game.Debug.Levels.INFORMATION
import com.one_of_many_simons.one_button_game.Libraries.collisions
import com.one_of_many_simons.one_button_game.Libraries.graphics
import com.one_of_many_simons.one_button_game.Libraries.listeners
import com.one_of_many_simons.one_button_game.dataClasses.Position
import com.one_of_many_simons.one_button_game.listeners.RefreshListener
import com.one_of_many_simons.one_button_game.Libraries.pathFinding
import com.one_of_many_simons.one_button_game.entities.player.Player
import com.one_of_many_simons.one_button_game.graphics.Graphics.Companion.DECOR_LAYER
import com.one_of_many_simons.one_button_game.graphics.Icons

class Projectile(position: Position, target: Position) : RefreshListener {
    private var position: Position
    private var directionIndex: Int
    private var icon = Icons.Interactive.projectile
    private var damage: Int = 5

    init {
        debug(PROJECTILE, CORE, "--- [Projectile.constructor]")

        this.position = position
        directionIndex = pathFinding.toPoint(position, target)

        onRefresh()
    }

    override fun onRefresh() {
        debug(PROJECTILE, CORE, ">>> [Projectile.onRefresh]")

        if (position.equals(Data.Player.position)) {
            Player.getDamage(damage)
            listeners.removeRefreshListener(this)
            return
        }

        position = pathFinding.getNextPosition(position, directionIndex)

        val couldMove = collisions.checkForCollision(graphics.getTile(position))
        if (couldMove == collisions.immovable) {
            listeners.removeRefreshListener(this)

            debug(PROJECTILE, CORE, "<<< [Projectile.onRefresh]")
            return
        }

        if (position.equals(Data.Player.position)) {
            Player.getDamage(damage)
            listeners.removeRefreshListener(this)

            debug(PROJECTILE, CORE, "<<< [Projectile.onRefresh]")
            return
        }

        graphics.drawTile(position, icon, DECOR_LAYER)

        debug(PROJECTILE, CORE, "<<< [Projectile.onRefresh]")
    }

    override fun getPosition(): Position {
        debug(PROJECTILE, INFORMATION, "--- [Projectile.getPosition]")

        return position
    }
}